R/expectation.r

```diff
@@ -11,20 +11,73 @@
 #' @keywords internal
 #' @export
 expectation <- function(passed, failure_msg, success_msg = "unknown", srcref = NULL) {
-  structure(
+  new_expectation(passed = passed, failure_msg = failure_msg,
+                  success_msg = success_msg, srcref = srcref)
+}
+
+new_expectation <- function(failure_msg, success_msg, srcref, ...,
+                            passed = FALSE, error = FALSE, skipped = FALSE) {
+  if (passed) {
+    class = c("expectation", "condition")
+  } else {
+    class = c("expectation", "error", "condition")
+  }
+
+  exp <- structure(
     list(
       passed = passed,
-      error = FALSE,
-      skipped = FALSE,
+      error = error,
+      skipped = skipped,
       failure_msg = failure_msg,
-      success_msg = success_msg,
-      srcref = srcref
+      success_msg = success_msg
     ),
-    class = "expectation"
+    class = class
   )
+
+  update_expectation(exp, srcref)
 }
 
-expectation_error <- function(error, calls) {
+update_expectation <- function(exp, srcref, info = NULL, label = NULL) {
+  exp$srcref <- srcref
+
+  if (!is.null(label)) {
+    exp$failure_msg <- paste0(label, " ", exp$failure_msg)
+    exp$success_msg <- paste0(label, " ", exp$success_msg)
+  }
+
+  if (!is.null(info)) {
+    exp$failure_msg <- paste0(exp$failure_msg, "\n", info)
+    exp$success_msg <- paste0(exp$success_msg, "\n", info)
+  }
+
+  exp$message <- if (exp$passed) exp$success_msg else exp$failure_msg
+
+  exp
+}
+
+
+as.expectation <- function(x, ...) UseMethod("as.expectation", x)
+
+#' @export
+as.expectation.default <- function(x, ...) {
+  stop("Don't know how to convert '", paste(class(x), collapse = "', '"),
+       "' to expectation.", call. = FALSE)
+}
+
+#' @export
+as.expectation.expectation <- function(x, ...) x
+
+#' @export
+as.expectation.logical <- function(x, message, ...) {
+  expectation(passed = x, failure_msg = message, srcref = find_test_srcref())
+}
+
+#' @export
+as.expectation.error <- function(x, ...) {
+  error <- x$message
+  calls <- x$calls
+  srcref <- x$srcref
```

Comment:
The srcref handling needs to be improved. This is currently just a placeholder, should be filled later by the calling handler to support file+line reporting.